### PR TITLE
V3.3.15

### PR DESCRIPTION
**Pull request by emveepee**
Initial changes for NextPVR v5 (subtitle, season, episode, disable timeshift).

### CODE REVIEW

**Comment by sub3** [comment]
Thanks Martin!

File: src/pvrclient-nextpvr.cpp [hunk 1]
```diff
@@ -279,6 +279,14 @@ bool cPVRClientNextPVR::Connect()
                     XBMC->QueueNotification(QUEUE_ERROR, XBMC->GetLocalizedString(30051), NEXTPVRC_MIN_VERSION_STRING);
                     return false;
                   }
+                  else if (version >= 50000)
+                  {
+                     if ( g_livestreamingmethod != RealTime)
+                     {
+                       g_livestreamingmethod = RealTime;
+                      XBMC->QueueNotification(QUEUE_ERROR,"v5 timeshifting disabled");
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:287]: Wrong indentation.

File: src/pvrclient-nextpvr.cpp [hunk 2]
```diff
@@ -279,6 +279,14 @@ bool cPVRClientNextPVR::Connect()
                     XBMC->QueueNotification(QUEUE_ERROR, XBMC->GetLocalizedString(30051), NEXTPVRC_MIN_VERSION_STRING);
                     return false;
                   }
+                  else if (version >= 50000)
+                  {
+                     if ( g_livestreamingmethod != RealTime)
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:284]: Blank between ( and  g.

File: src/pvrclient-nextpvr.cpp [hunk 3]
```diff
@@ -563,17 +572,44 @@ PVR_ERROR cPVRClientNextPVR::GetEpg(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
         memset(&broadcast, 0, sizeof(EPG_TAG));
 
-        char title[128];
-        char description[1024];
-
-        strncpy(title, pListingNode->FirstChildElement("name")->FirstChild()->Value(), sizeof title);
+        string title;
+        string description;
+        string subtitle;
+        title = pListingNode->FirstChildElement("name")->FirstChild()->Value();
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:578]: This is not  C. Why don't you declare and init title in one line?

File: src/pvrclient-nextpvr.cpp [hunk 4]
```diff
@@ -563,17 +572,44 @@ PVR_ERROR cPVRClientNextPVR::GetEpg(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
         if (pListingNode->FirstChildElement("description") != NULL && pListingNode->FirstChildElement("description")->FirstChild() != NULL)
         {
-          PVR_STRCPY(description, pListingNode->FirstChildElement("description")->FirstChild()->Value());
+          description = pListingNode->FirstChildElement("description")->FirstChild()->Value();
+        }
+        else
+        {
+          description = "";
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:585]: Isn't description always empty here?

File: src/pvrclient-nextpvr.cpp [hunk 5]
```diff
@@ -563,17 +572,44 @@ PVR_ERROR cPVRClientNextPVR::GetEpg(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
+        }
+
+        if (pListingNode->FirstChildElement("subtitle") != NULL && pListingNode->FirstChildElement("subtitle")->FirstChild() != NULL)
+        {
+          subtitle = pListingNode->FirstChildElement("subtitle")->FirstChild()->Value();
+          if (description == subtitle + ":")
+          {
+            description = "";
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:593]: description.clear() is more efficient than assigning empty string.

File: src/pvrclient-nextpvr.cpp [hunk 6]
```diff
@@ -563,17 +572,44 @@ PVR_ERROR cPVRClientNextPVR::GetEpg(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
+          }
+          else if (StringUtils::StartsWith(description,subtitle + ": "))
+          {
+            description = description.substr(subtitle.length()+2);
+          }
         }
         else
         {
-          description[0] = '\0';
+          subtitle = "";
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:602]: subtitle.clear()

File: src/pvrclient-nextpvr.cpp [hunk 7]
```diff
@@ -563,17 +572,44 @@ PVR_ERROR cPVRClientNextPVR::GetEpg(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
+        }
+
+        if (pListingNode->FirstChildElement("year") != NULL && pListingNode->FirstChildElement("year")->FirstChild() != NULL)
```
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp:605]: in c++, nullptr should be used instead of NULL

**Comment by ksooo** [comment]
Let me know if you guys don't care about "stylistic" code enhancements, then I will  no longer look over your PRs. Personally, I think the last PRs could have gotten some more love before merge.

**Comment by sub3** [comment]
Thanks ksooo - even this stylistic feedback is appreciated.  Sorry it wasn't a cleaner PR from us. When we're focusing on application function, some times we forget this type of stuff. We will work through your feedback items, and address with the next PR. Cheers.

**Comment by emveepee** [comment]
Yes  I was looking for feedback, the PR got tagged a bit earlier than I expected.   I do want to move more of the old C code to C++ as I go so don't know all the conventions.   Things like NULL to nullptr will be more for a style cleanup PR.

The reason for some of the odd looking code is the API from NextPVR supports multiple clients beyond Kodi and sub didn't want to bump the server API version.  Currently the description is a combination of  "episode name: description."  Now that the episode name is coming in a separate field in v5, I have to parse it with the colon from the description.  This means at times there will be no description.

**Comment by ksooo** [comment]
>  I do want to move more of the old C code to C++ as I go so don't know all the conventions. Things like NULL to nullptr will be more for a style cleanup PR.

 The second sentence contradicts the first. Either you want to make the changes as you go or with dedicated cleanup commits. ;-)

**Comment by emveepee** [comment]
All the the other XML parsing code uses NULL, I thought it would be odd changing them one at a time but if that is better style I will do that in the future.

**Comment by ksooo** [comment]
> All the the other XML parsing code uses NULL, I thought it would be odd changing them one at a time 

 There is no right or wrong here. For Kodi, we fix coding guideline violations whenever we touch a code line (this is how I take "as we go"). Addons can handle this differently.

 I just commented where I thought the code could be improved. It's completely up to you what you make out of this.

**Comment by emveepee** [comment]
Excellent comment on fixing the code as we touch the line so I will do that going forward.  Much appreciated once again.

**Pull request merged**